packages/hint-compat-api/src/utils/parse-supports.ts

```diff
@@ -0,0 +1,184 @@
+                    }
+                    tokens.push(temp);
+                    temp = '';
+                }
+                tokens.push(char);
+                continue;
+            case ':':
+                /*
+                 * A colon switches to the declaration value parsing state.
+                 * The in-progress temp token is the declaration name.
+                 * Pass `i + 1` to skip over the colon when parsing the value.
+                 */
+                [i, token] = tokenizeDeclarationValue(params, temp, i + 1);
+                i -= 1; // -1 to process ending ')' token on next iteration.
+                temp = '';
+                tokens.push(token);
+                break;
+            default:
+                // Otherwise consider part of an in-progress temp token.
+                temp += char;
+        }
+    }
+
+    // Remove redundant `(` and `)` tokens around `Declaration`s before returning.
+    return tokens.filter((token, i) => {
+        return !(
+            (token === '(' && isDeclaration(tokens[i + 1])) ||
+            (token === ')' && isDeclaration(tokens[i - 1]))
+        );
+    });
+};
+
+const parseParams = (params: string): DeclarationGroup => {
+    const tokens = tokenizeParams(params);
+
+    const groups: DeclarationGroup[] = [{
+        nodes: [],
+        type: 'and'
+    }];
+
+    for (const token of tokens) {
+        switch (token) {
+            case '(':
+                // Left paren creates a new current group.
+                groups.unshift({
+                    nodes: [],
+                    type: 'and'
+                });
+                // New group is also a node in the previous group.
+                groups[1].nodes.push(groups[0]);
+                break;
+            case ')':
+                // Right paren closes current group.
+                groups.shift();
+                break;
+            case 'and':
+            case 'or':
+            case 'not':
+                // Operator changes current group type.
+                groups[0].type = token;
+                break;
+            default:
+                // Declarations are nodes in the current group.
+                groups[0].nodes.push(token);
+        }
+    }
+
+    if (groups.length !== 1) {
+        throw new Error('Mis-matched parenthesis');
+    }
+
+    return groups[0];
+};
+
+/**
+ * Parse an `@supports` params string into an AST.
+ *
+ * ```css
+ * @supports not ((display: flex) or (display: grid))
+ * ```
+ *
+ * ```js
+ * {
+ *     type: 'not',
+ *     nodes: [
+ *         {
+ *             type: 'or',
+ *             nodes: [
+ *                 { prop: 'display', value: 'flex' },
+ *                 { prop: 'display', value: 'grid' }
+ *             ]
+ *         }
+ *     ]
+ * }
+ * ```
+ */
+export const parseSupports = (params: string): DeclarationGroup | null => {
+    try {
+        return parseParams(params);
+    } catch (e) {
+        return null;
```

Comment:
Looking at this again I'm thinking I should throw a specific type of error for known invalid states and only filter errors of that type out here. This would avoid silencing unexpected bugs is the parser...